Xi-syntect panicking in background but no client crash

Im consistently able to recreate this issue but only with a huge file with ~110K lines. To recreate the issue:
- Open the file and select all. 
- Hit backspace to empty the file. 
- Save the file. 

If you look at the log file you see stacktrace similar to 
```
stylemap can't resolve 8 <-- thousands of line 
thread 'main' panicked at 'index 1845405 out of range for slice of length 1048576', libcore/slice/mod.rs:2334:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::continue_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::slice::slice_index_len_fail
   9: <core::ops::range::Range<usize> as core::slice::SliceIndex<[T]>>::index
  10: core::slice::<impl core::ops::index::Index<I> for [T]>::index
  11: xi_plugin_lib::base_cache::ChunkCache::simple_delete
  12: xi_plugin_lib::base_cache::ChunkCache::should_clear
  13: <xi_plugin_lib::base_cache::ChunkCache as xi_plugin_lib::Cache>::update
  14: <xi_plugin_lib::state_cache::StateCache<S> as xi_plugin_lib::Cache>::update
  15: <xi_plugin_lib::view::View<C>>::update
  16: <xi_plugin_lib::dispatch::Dispatcher<'a, P>>::do_update
  17: <xi_plugin_lib::dispatch::Dispatcher<'a, P> as xi_rpc::Handler>::handle_request
  18: <xi_rpc::RpcLoop<W>>::mainloop::{{closure}}
  19: crossbeam_utils::thread::scope::{{closure}}
  20: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  21: std::panicking::try::do_call
  22: __rust_maybe_catch_panic
  23: std::panicking::try
  24: std::panic::catch_unwind
  25: crossbeam_utils::thread::scope
  26: <xi_rpc::RpcLoop<W>>::mainloop
  27: xi_plugin_lib::mainloop
  28: xi_syntect_plugin::main
  29: std::rt::lang_start::{{closure}}
  30: std::panicking::try::do_call
  31: __rust_maybe_catch_panic
  32: std::rt::lang_start_internal
  33: std::rt::lang_start
  34: main
[2019-03-13][20:52:05][xi_core_lib::event_context][ERROR] plugin shutdown, do something PeerDisconnect
[2019-03-13][20:52:05][xi_core_lib::tabs][WARN] plugin PluginPid(1) exited with result Ok(())
client: plugin stopped xi-syntect-plugin
[2019-03-13][20:52:05][xi_core_lib::tabs][DEBUG] Event in open file NoticeRemove("/sujitjoshi/notify/append")
```

Im using latest of xi-core and xi-mac.

--- a/rust/plugin-lib/src/base_cache.rs
+++ b/rust/plugin-lib/src/base_cache.rs
@@ -333,7 +333,7 @@ impl ChunkCache {
         if start < self.offset || start > self.offset + self.contents.len() {
             true
         } else if delta.is_simple_delete() {
-            self.simple_delete(start, end);
+            self.simple_delete(start, std::cmp::min(end, self.offset + self.contents.len()));
             false
         } else if let Some(text) = delta.as_simple_insert() {
             assert_eq!(iv.size(), 0);
@@ -701,6 +701,30 @@ mod tests {
         c.simple_delete(start, end);
 
         assert_eq!(&c.line_offsets, &[7, 11]);
+    }
+
+    #[test]
+    fn large_delete() {
+        // Issue #1136 on github
+        let large_str = "This string literal is larger than CHUNK_SIZE.";
+        assert!(large_str.len() > CHUNK_SIZE);
+
+        let data = GetDataResponse {
+            // Emulate a cache that has only a part of the buffer.
+            chunk: large_str.split_at(CHUNK_SIZE).0.into(),
+            offset: 0,
+            first_line: 0,
+            first_line_offset: 0,
+        };
+        let mut c = ChunkCache::default();
+        c.reset_chunk(data);
+
+        // This delta deletes everything.
+        let delta =
+            Delta::simple_edit(Interval::new(0, large_str.len()), "".into(), large_str.len());
+        assert!(delta.is_simple_delete());
+
+        c.update(Some(&delta), delta.new_document_len(), 1, 1); // Should succeed
     }
 
     #[test]

--- a/rust/plugin-lib/src/base_cache.rs
+++ b/rust/plugin-lib/src/base_cache.rs
@@ -333,7 +333,10 @@ impl ChunkCache {
         if start < self.offset || start > self.offset + self.contents.len() {
             true
         } else if delta.is_simple_delete() {
-            self.simple_delete(start, std::cmp::min(end, self.offset + self.contents.len()));
+            // Don't go over cache boundary.
+            let end = end.min(self.offset + self.contents.len());
+            
+            self.simple_delete(start, end);
             false
         } else if let Some(text) = delta.as_simple_insert() {
             assert_eq!(iv.size(), 0);

--- a/rust/plugin-lib/src/base_cache.rs
+++ b/rust/plugin-lib/src/base_cache.rs
@@ -335,7 +335,7 @@ impl ChunkCache {
         } else if delta.is_simple_delete() {
             // Don't go over cache boundary.
             let end = end.min(self.offset + self.contents.len());
-            
+
             self.simple_delete(start, end);
             false
         } else if let Some(text) = delta.as_simple_insert() {

